src/mame/drivers/uknc.cpp

```diff
@@ -9,97 +9,946 @@ Graphics 640x288 pixels.
+// map 3
+	map(0300000, 0317777).rom().region("subcpu", 0).bankw("plane0bank4");
+}
+
+/*
+ *	bits 0-2 of map number = value of bits 5-7 in BCSR (ROM banks replacement: 0 - writes ignored, 1 - writes go to VRAM plane 0)
+ *	  5		120000..137777
+ *	  6		140000..157777
+ *	  7		160000..177777
+ */
+void uknc_state::uknc_rom_mem(address_map &map)
+{
+// map 0: all ROM, no VRAM
+	map(0000000, 0057777).rom().region("subcpu", 020000);
+// map 1
+	map(0100000, 0117777).rom().region("subcpu", 020000).bankw("plane0bank5");
+	map(0120000, 0157777).rom().region("subcpu", 040000);
+// map 2
+	map(0200000, 0217777).rom().region("subcpu", 020000);
+	map(0220000, 0237777).rom().region("subcpu", 040000).bankw("plane0bank6");
+	map(0240000, 0257777).rom().region("subcpu", 060000);
+// map 3
+	map(0300000, 0317777).rom().region("subcpu", 020000).bankw("plane0bank5");
+	map(0320000, 0337777).rom().region("subcpu", 040000).bankw("plane0bank6");
+	map(0340000, 0357777).rom().region("subcpu", 060000);
+// map 4
+	map(0400000, 0437777).rom().region("subcpu", 020000);
+	map(0440000, 0457777).rom().region("subcpu", 060000).bankw("plane0bank7");
+// map 5
+	map(0500000, 0517777).rom().region("subcpu", 020000).bankw("plane0bank5");
+	map(0520000, 0537777).rom().region("subcpu", 040000);
+	map(0540000, 0557777).rom().region("subcpu", 060000).bankw("plane0bank7");
+// map 6
+	map(0600000, 0617777).rom().region("subcpu", 020000);
+	map(0620000, 0637777).rom().region("subcpu", 040000).bankw("plane0bank6");
+	map(0640000, 0657777).rom().region("subcpu", 060000).bankw("plane0bank7");
+// map 7
+	map(0700000, 0717777).rom().region("subcpu", 020000).bankw("plane0bank5");
+	map(0720000, 0737777).rom().region("subcpu", 040000).bankw("plane0bank6");
+	map(0740000, 0757777).rom().region("subcpu", 060000).bankw("plane0bank7");
 }
 
+static const z80_daisy_config sub_daisy_chain[] =
+{
+//	{ "event" },
+//	{ "timer" },
+//	{ "keyboard" },
+//	{ "reset" },
+//	{ "subchan" },
+	{ "cart" },
+	{ nullptr }
+};
+
 /* Input ports */
+
+// some models had R and G outputs swapped and others had no Y (intensity) output
 static INPUT_PORTS_START( uknc )
+	PORT_START("yrgb")
+	PORT_DIPNAME(0x03, 0x00, "variant")
+	PORT_DIPSETTING(0x00, "Color, YRGB")
+	PORT_DIPSETTING(0x01, "Color, YGRB")
+	PORT_DIPSETTING(0x02, "Color, xRGB")
+	PORT_DIPSETTING(0x03, "Grayscale")
 INPUT_PORTS_END
 
 
 void uknc_state::machine_reset()
 {
+	bcsr_w(0, 01401, 0xffff);
+	scsr_w(0, 0, 0xffff);
 }
 
 void uknc_state::machine_start()
 {
+	// 3 bitplanes
+	m_videoram = std::make_unique<uint8_t[]>(32768 * 3);
+
+	membank("plane0bank4")->set_base(&m_videoram[0]);
+	membank("plane0bank5")->set_base(&m_videoram[020000]);
+	membank("plane0bank6")->set_base(&m_videoram[040000]);
+	membank("plane0bank7")->set_base(&m_videoram[060000]);
+}
+
+void uknc_state::video_start()
+{
+	m_scanlines = std::make_unique<scanline_t[]>(UKNC_DISP_VERT);
+
+	m_tmpbmp.allocate(UKNC_DISP_HORZ, UKNC_DISP_VERT);
+}
```

Comment:
Avoid using `video_start` if it’s not necessary.  It’s a hold-over from the pre-device model world.  There’s no reason not to put this in `machine_start`.

Reply:
changed.